node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+//
+// When we're choosing bitfields to include, the rule should be simple:
+// maximize availability. So basically, include all bitfields. And then
+// choose a coherent set of candidates along with that.
+async fn send_inherent_data(
+	relay_parent: Hash,
+	signed_bitfields: Vec<SignedAvailabilityBitfield>,
+	mut backed_candidates: Vec<BackedCandidate>,
+	return_sender: oneshot::Sender<ProvisionerInherentData>,
+	mut from_job: mpsc::Sender<FromJob>,
+) -> Result<(), Error> {
+	let availability_cores = match request_availability_cores(relay_parent, &mut from_job)
+		.await?
+		.await?
+	{
+		Ok(cores) => cores,
+		Err(runtime_err) => {
+			// Don't take down the node on runtime API errors.
+			log::warn!(target: "provisioner", "Encountered a runtime API error: {:?}", runtime_err);
+			return Ok(());
+		}
+	};
+
+	// availability cores indexed by their para_id for ease of use
+	let cores_by_id: HashMap<_, _> = availability_cores
+		.iter()
+		.enumerate()
+		.filter_map(|(idx, core_state)| Some((core_state.para_id()?, (idx, core_state))))
+		.collect();
+
+	// utility
+	let mut parachains_represented = HashSet::new();
+
+	// coherent candidates fulfill these conditions:
+	//
+	// - only one per parachain
+	// - any of:
+	//   - this para is assigned to a `Scheduled` core
+	//   - this para is assigned to an `Occupied` core, and any of:
+	//     - it is `next_up_on_available` and the bitfields we are including, merged with
+	//       the `availability` vec, form 2/3+ of validators
+	//     - it is `next_up_on_time_out` and the bitfields we are including, merged with
+	//       the `availability_ vec, for <2/3 of validators, and `time_out_at` is
+	//       the block we are building.
+	//
+	// postcondition: they are sorted by core index
+	backed_candidates.retain(|candidate| {
+		// only allow the first candidate per parachain
+		let para_id = candidate.candidate.descriptor.para_id;
+		if !parachains_represented.insert(para_id) {
+			return false;
+		}
+
+		let (core_idx, core) = match cores_by_id.get(&para_id) {
+			Some(core) => core,
+			None => return false,
+		};
+
+		match core {
+			CoreState::Free => false,
+			CoreState::Scheduled(_) => true,
+			CoreState::Occupied(occupied) => {
+				if let Some(ref scheduled) = occupied.next_up_on_available {
+					return scheduled.para_id == para_id
+						&& merged_bitfields_are_gte_two_thirds(
+							*core_idx,
+							&signed_bitfields,
+							&occupied.availability,
+						);
+				}
+				if let Some(ref scheduled) = occupied.next_up_on_time_out {
+					return scheduled.para_id == para_id
+						// TODO: this function currently doesn't know about what block we're building
+						//   get it from the runtime? where?
+						&& unimplemented!("occupied.time_out_at == the block we are building")
+						// REVIEW: the guide says we have to ensure that the merged bitfields are < 2/3 in this case,
+						// but intuition suggests that we'd still want it to be >=. Just want a double-check that
```

Comment:
Take a look at the `Inclusion` module of the runtime. It processes bitfields before time-outs. So if the bitfields would make the previous thing occupying the core available, we would want to use the thing scheduled in `next_up_on_available`, which is handled in the branch above. 

The guidance I gave to check that the bitfields we're including do _not_ make the thing available is correct because that is the only way the time-out gets processed, and we are trying to include the `next_up_on_time_out`.